src/treetensornetworks/opsum_to_ttn/opsum_to_ttn.jl

```diff
@@ -88,130 +77,119 @@ function pos_in_link!(linkmap::Dict, k)
 
   # Build compressed finite state machine representation (symbolic_ttn)
-  for v in ordered_verts
+  for v in vertices(sites)
+    symbolic_ttn[v] = vertex_data_eltype()
     v_degree = degree(sites, v)
+
     # For every vertex, find all edges that contain this vertex
     # (align_and_reorder_edges makes the output of indicident edges match the
     #  direction and ordering match that of ordered_edges)
-    edges = align_and_reorder_edges(incident_edges(sites, v), ordered_edges)
+    v_edges = align_and_reorder_edges(incident_edges(sites, v), ordered_edges)
 
     # Use the corresponding ordering as index order for tensor elements at this site
-    dim_in = findfirst(e -> dst(e) == v, edges)
-    edge_in = (isnothing(dim_in) ? nothing : edges[dim_in])
-    dims_out = findall(e -> src(e) == v, edges)
-    edges_out = edges[dims_out]
+    dim_in = findfirst(e -> dst(e) == v, v_edges)
+    edge_in = (isnothing(dim_in) ? nothing : v_edges[dim_in])
+    dims_out = findall(e -> src(e) == v, v_edges)
+    edges_out = v_edges[dims_out]
 
     # For every site v' except v, determine the incident edge to v
     # that lies in the edge_path(v',v)
     # TODO: better way to make which_incident_edge below?
+    # TODO: instead of split at vertex, should
+    #       be able to traverse subtrees formed by cutting incoming bond
     subgraphs = split_at_vertex(sites, v)
-    _boundary_edges = [
-      only(boundary_edges(underlying_graph(sites), subgraph)) for subgraph in subgraphs
-    ]
-    _boundary_edges = align_edges(_boundary_edges, edges)
+    _boundary_edges = [only(boundary_edges(sites, subgraph)) for subgraph in subgraphs]
+    _boundary_edges = align_edges(_boundary_edges, v_edges)
     which_incident_edge = Dict(
       Iterators.flatten([
         subgraphs[i] .=> ((_boundary_edges[i]),) for i in eachindex(subgraphs)
       ]),
     )
 
-    # Sanity check, leaves only have single incoming or outgoing edge
-    @assert !isempty(dims_out) || !isnothing(dim_in)
-    (isempty(dims_out) || isnothing(dim_in)) && @assert is_leaf_vertex(sites, v)
-
     for term in opsum
-      # Loop over OpSum and pick out terms that act on current vertex
-      ops = ITensors.terms(term)
-      if v in ITensors.site.(ops)
-        crosses_vertex = true
-      else
-        crosses_vertex =
-          !isone(length(Set([which_incident_edge[site] for site in site.(ops)])))
-      end
-      # If term doesn't cross vertex, skip it
-      crosses_vertex || continue
+      ops = terms(term)
+      op_sites = ITensors.sites(term)
+
+      # Only consider terms crossing current vertex
+      crosses_v((s1, s2),) = (v ∈ vertex_path(sites, s1, s2))
+      v_crossed = (v ∈ op_sites) || any(crosses_v, partition(op_sites, 2, 1))
+      # TODO: seems we could make this just
+      # v_crossed = (v ∈ vertices(steiner_tree(sites,op_sites))))
+      # but it is not working - steiner_tree seems to have spurious extra vertices?
```

Comment:
Can we define our own `tree_steiner_tree` function that is specific to tree graphs and calls `a_star` between each pair of specified vertices?

Reply:
I believe @b-kloss also reported the `steiner_tree` didn't always give the minimal Steiner tree, even on trees. Probably `Graphs.jl` should have a special code path if the graph is a tree, that would make for a good PR. Also reporting an issue to `Graphs.jl` of an example where `steiner_tree` fails would be a good start.

Reply:
Please check whether this is a performance regression. The previous code was an optimization to remove a significant performance bottleneck when using `vertex_path` logic. Note that the test cases in ITensorNetworks/test are not sufficiently hard to verify this. The bottleneck was evident for multi-orbital impurity problems with star-geometry on a fork graph with O(100-1000) sites on each leg of the fork.

Reply:
I see `NamedGraphs.GraphsExtensions.vertex_path` may not be implemented be implemented in the best way: https://github.com/mtfishman/NamedGraphs.jl/blob/v0.6.0/src/lib/GraphsExtensions/src/abstractgraph.jl#L436-L449. Is it possible that optimizing that (say by using `a_star`) might fix the performance issue you were seeing on large graphs?

Additionally I wonder if we can rewrite the code in a nice way using `vertex_path`/`steiner_tree` and then fix any performance issues with a general solution like memoization to cache paths that have already been computed, i.e. find a way to get the best of both worlds in terms of simple code and good performance.

Reply:
Concretely, I think the previous code was O(n_vertices^2 + ...) and O(n_terms) * O(average__n_site_per_term)).

Reply:
I doubt that `a_star` is the way to go for trees. I think there should be an algorithm that takes a set of vertices on a tree and returns the span of said vertices (i.e. union of all possible vertex_paths among the set) that has cost O(vertices_in_span). Basically the vertex_path falls out directly from using a tree datastructure, you pick a root and then walk the tree towards the root from each vertex in the pair until you either hit the other vertex or a joint parent. You put the visited vertices in a dictionary, so that lookup is O(1) and the algorithm should overall scale as O(span). Since we are only interested in whether a given vertex is within the span, we can terminate once the site is found in the span, making the cost O(1)-O(span).
I imagine if this was efficiently implemented, the prefactors become small enough that it shouldn't be noticeably slower than the previous approach for any realistic Hamiltonian.

Reply:
Thanks. So to summarize your comment, if we had an implementation of `steiner_tree(graph, vertices)` that scaled at worst as `O(nv(steiner_tree(graph, vertices)))` if `is_tree(graph)`, then used that function here, that could help to make the code simpler and also likely get good enough efficiency in the large tree limit.

I agree that my suggestion of repeatedly using `a_star` between the specified vertices may not be the absolute best way to implement it, I was more suggesting that as a simple first thing to try. The algorithm you suggest sounds good.

Reply:
Yes, that summarizes it.
However, this PR still uses _which_incident_edge, so it already constructed an object that allows to determine crosses_vertex at cost O(n_sites_in_term).

Reply:
Our goal now is more about simplifying the code as much as possible, while keeping in mind parts that may be bottlenecks and need to be optimized, and then once the code is simpler we will go back and optimize. It's part of a bigger rewrite based on https://github.com/ITensor/ITensorNetworks.jl/issues/117.

Reply:
But it is helpful to hear about parts you found to be bottlenecks if they are not written properly.

Reply:
Thanks, Benedikt. That's really helpful background information as well as feedback about which graph algorithms we might want to use versus avoid. I'm on a learning curve with some of the graph mathematics. Based on this discussion, I'm going to create some performance benchmarks for me to test over time (is there already a collection of these somewhere?) and to check against the version before the previous PR. I definitely believe you that some of these changes are performance regressions so we plan to get that performance back as we iterate the design.